File: 2023/day3/part2.py
def get_nbr(lines, i, j):
	n = ''
	while j > 0 and lines[i][j].isdigit():
		j-=1
	if not lines[i][j].isdigit():
		j+=1
	start = j
	while j < len(lines[i]) and lines[i][j].isdigit():
		n += lines[i][j]
		j+=1
	end = j - 1
	#print(n, start, end)
	return (int(n), start, end)

def main():
	lines = []
	with open("input.txt") as f:
		lines = f.read().split('\n')[:-1]

	#lines = ['467..114..', '...*......', '..35..633.', '......#...', '617*......', '.....+.58.', '..592.....', '......755.', '...$.....*', '.664...598']


	sm = 0

	for i in range(len(lines)):
		line = lines[i]
		for j in range(len(line)):
			if line[j] == '*':
				lst = []
				if (i > 0 and lines[i-1][j].isdigit()):
					lst.append(get_nbr(lines, i-1, j))
				if (i > 0 and j > 0 and lines[i-1][j-1].isdigit()):
					lst.append(get_nbr(lines, i-1, j-1))
				if (i > 0 and j < len(line) - 1 and lines[i-1][j+1].isdigit()):
					lst.append(get_nbr(lines, i-1, j+1))
				if (j > 0 and lines[i][j-1].isdigit()):
					lst.append(get_nbr(lines, i, j-1))
				if (j < len(line) - 1 and lines[i][j+1].isdigit()):
					lst.append(get_nbr(lines, i, j+1))
				if (i < len(lines) - 1 and lines[i+1][j].isdigit()):
					lst.append(get_nbr(lines, i+1, j))
				if (i < len(lines) - 1 and j > 0 and lines[i+1][j-1].isdigit()):
					lst.append(get_nbr(lines, i+1, j-1))
				if (i < len(lines) - 1 and j < len(line) - 1 and lines[i+1][j+1].isdigit()):
					lst.append(get_nbr(lines, i+1, j+1))
				lst = list(set(lst))
				if len(lst) == 2:
					#print(lst)
					sm += lst[0][0] * lst[1][0]

	print(sm)

main()

File: 2024/day2/part2.cpp
#include <vector>
#include <string>
#include <ctype.h>
#include <cstdlib>
#include <fstream>
#include <iostream>

int main()
{
    size_t i;
    size_t j;
    int     count;

    bool	inc;
    bool    removed;
    std::string nbr;
    std::string line;
    std::vector<int> vec;

    std::ifstream input("input.txt");

    if (!input.is_open())
    {
        std::cout << "cannot open input.txt" << std::endl;
        return (1);
    }

    count = 0;
    while (getline(input, line))
    {
		if (line.empty())
			continue ;

        i = 0;
        while (line[i])
        {
            j = i;
            while (line[i] && isdigit(line[i]))
                i++;

            if (j != i)
			{
				nbr = line.substr(j, i - j);
	            vec.push_back(atoi(nbr.c_str()));
			}
			if (line[i])
				i++;
        }

        inc = vec[0] < vec[1];

        i = 1;
        count++;
        removed = false;
        while (i < vec.size())
        {
			j = abs(vec[i] - vec[i - 1]);
			if ((vec[i - 1] < vec[i]) != inc || ((i + 1) < vec.size() && (vec[i] < vec[i + 1]) != inc) || !(j >= 1 && j <= 3))
			{
                std::cout << "i " << i << std::endl;
                if (removed)
                {
                    std::cout << std::endl << line << std::endl;
                    count--;
                    break ;
                }
                removed = true;
			}
            i++;
        }
        vec.clear();
    }

	std::cout << "safe reports: " << count << std::endl;

    input.close();
}
